analyzers/src/SonarAnalyzer.CSharp/SymbolicExecution/Roslyn/PublicMethodArgumentsShouldBeCheckedForNull.cs

```diff
@@ -18,16 +18,55 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
 namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks.CSharp;
 
 public class PublicMethodArgumentsShouldBeCheckedForNull : SymbolicRuleCheck
 {
     private const string DiagnosticId = "S3900";
-    private const string MessageFormat = "Refactor this method to add validation of parameter '{0}' before using it.";
+    private const string MessageFormat = "{0}";
 
     internal static readonly DiagnosticDescriptor S3900 = DescriptorFactory.Create(DiagnosticId, MessageFormat);
 
     protected override DiagnosticDescriptor Rule => S3900;
 
-    public override bool ShouldExecute() => false;
+    public override bool ShouldExecute() => true;
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        if (context.Operation.Instance.Kind == OperationKindEx.ParameterReference
+            && context.Operation.Instance.ToParameterReference() is { WrappedOperation: var reference, WrappedOperation.Syntax: var syntax, Parameter: var parameterSymbol }
+            && !parameterSymbol.Type.IsValueType
+            && IsParameterDereferenced(context.Operation)
+            && !IgnoreBecauseOfParameterAttribute(parameterSymbol)
+            && NullableStateIsNotKnownForParameter(parameterSymbol)
+            && SemanticModel.GetDeclaredSymbol(Node) is { } methodSymbol)
+        {
+            var message = methodSymbol.IsConstructor()
+                ? "Refactor this constructor to avoid using members of parameter '{0}' because it could be null."
+                : "Refactor this method to add validation of parameter '{0}' before using it.";
+            ReportIssue(reference, string.Format(message, syntax.ToString()), context);
+        }
+
+        return context.State;
+
+        static bool IsParameterDereferenced(IOperationWrapperSonar operation) =>
+            operation.Parent != null
+            && operation.Parent.IsAnyKind(
+                OperationKindEx.Invocation,
+                OperationKindEx.FieldReference,
+                OperationKindEx.PropertyReference,
+                OperationKindEx.EventReference,
+                OperationKindEx.Await,
+                OperationKindEx.ArrayElementReference);
+
+        bool IgnoreBecauseOfParameterAttribute(IParameterSymbol symbol) =>
+            symbol.HasAttribute(KnownType.Microsoft_AspNetCore_Mvc_FromServicesAttribute);
+
+        bool NullableStateIsNotKnownForParameter(IParameterSymbol symbol) =>
+            !context.HasConstraint(symbol, ObjectConstraint.Null)
+            && !context.HasConstraint(symbol, ObjectConstraint.NotNull)
+            && SemanticModel.GetTypeInfo(context.Operation.Instance.Syntax).Nullability().FlowState != NullableFlowState.NotNull;
```

Comment:
```suggestion
            && SemanticModel.GetTypeInfo(context.Operation.Instance.Syntax).Nullability().FlowState == NullableFlowState.None;
```

Reply:
Removed the flow state analysis for now, as there are no UTs for it. I'll add them in a separate PR.